sentry_sdk/integrations/opentelemetry/sampler.py

```diff
@@ -0,0 +1,105 @@
+from random import random
+
+from opentelemetry import trace
+from opentelemetry.sdk.trace.sampling import Sampler, SamplingResult, Decision
+from opentelemetry.trace import SpanKind
+from opentelemetry.util.types import Attributes
+
+import sentry_sdk
+from sentry_sdk.integrations.opentelemetry.consts import SENTRY_TRACE_STATE_DROPPED
+from sentry_sdk.tracing_utils import has_tracing_enabled
+from sentry_sdk.utils import is_valid_sample_rate, logger
+
+from typing import Optional, Sequence
+
+
+def get_parent_sampled(parent_context, trace_id):
+    # type: (Optional["Context"], int) -> Optional[bool]
+    is_span_context_valid = parent_context is not None and parent_context.is_valid
+
+    # Only inherit sample rate if `traceId` is the same
+    if is_span_context_valid and parent_context.trace_id == trace_id:
+        # this is getSamplingDecision in JS
+        if parent_context.trace_flags.sampled:
+            return True
+
+        dropped = parent_context.trace_state.get(SENTRY_TRACE_STATE_DROPPED) == "true"
+        if dropped:
+            return False
+
+        # TODO-anton: fall back to sampling decision in DSC (for this die DSC needs to be set in the trace_state)
+
+    return None
+
+
+class SentrySampler(Sampler):
+    def should_sample(
+        self,
+        parent_context: Optional["Context"],
+        trace_id: int,
+        name: str,
+        kind: Optional[SpanKind] = None,
+        attributes: Attributes = None,
+        links: Optional[Sequence["Link"]] = None,
+        trace_state: Optional["TraceState"] = None,
+    ) -> "SamplingResult":
+        client = sentry_sdk.get_client()
+
+        # No tracing enabled, thus no sampling
+        if not has_tracing_enabled(client.options):
+            return SamplingResult(Decision.DROP)
+
+        # Check if sampled=True was passed to start_transaction
+        # TODO-anton: Do we want to keep the start_transaction(sampled=True) thing?
+
+        parent_span = trace.get_current_span()
+        parent_context = parent_span and parent_span.get_span_context()
+
+        sample_rate = None
+        # Check if there is a traces_sampler
+        # Traces_sampler is responsible to check parent sampled to have full transactions.
+        has_traces_sampler = callable(client.options.get("traces_sampler"))
+        if has_traces_sampler:
+            # TODO-anton: Make proper sampling_context
+            sampling_context = {
+                "transaction_context": {
+                    "name": name,
+                },
+                "parent_sampled": parent_context.trace_flags.sampled,
+            }
+
+            sample_rate = client.options["traces_sampler"](sampling_context)
+
+        # Check if there is a parent with a sampling decision
+        if sample_rate is None:
```

Comment:
these need to if/elif to match with these semantics
```
        # we would have bailed already if neither `traces_sampler` nor
        # `traces_sample_rate` were defined, so one of these should work; prefer
        # the hook if so
        sample_rate = (
            client.options["traces_sampler"](sampling_context)
            if callable(client.options.get("traces_sampler"))
            else (
                # default inheritance behavior
                sampling_context["parent_sampled"]
                if sampling_context["parent_sampled"] is not None
                else client.options["traces_sample_rate"]
            )
        )
```
otherwise the behavior subtly changes, for instance when `traces_sampler` returns None, we do extra stuff now which shouldn't be done

Reply:
fixed